File: util/gen_pet_names.py
#!/usr/bin/env python

import coinach
import csv
import csv_util
import os

_OUTPUT_FILE = "pet_names.ts"


if __name__ == "__main__":
    keys = ["Name"]

    tables = {}
    for lang in ["en", "fr", "de", "ja"]:
        reader = coinach.CoinachReader(verbose=True)
        tables[lang] = csv_util.make_map(reader.exd("Pet", lang=lang), keys)
    tables["cn"] = csv_util.get_cn_table("Pet", keys)
    tables["ko"] = csv_util.get_ko_table("Pet", keys)

    tables = {lang: [name for name in table.keys() if name] for lang, table in tables.items()}

    writer = coinach.CoinachWriter(verbose=True)
    header = """import { Lang } from '../../resources/languages';

type PetData = {
  [name in Lang]: readonly string[];
};"""
    writer.writeTypeScript(
        filename=os.path.join("resources", _OUTPUT_FILE),
        scriptname=os.path.basename(os.path.abspath(__file__)),
        header=header,
        type="PetData",
        as_const=False,
        data=tables,
    )

File: util/gen_pet_names.ts
import path from 'path';

import fetch from 'node-fetch';

import { CoinachWriter } from './coinach';
import { getCnTable, getKoTable } from './csv_util';

const _OUTPUT_FILE = 'pet_names.ts';

const keys = ['Name'];
const localeKeys = [...keys, 'Name_de', 'Name_fr', 'Name_ja'];

type XivapiResult = {
  Results: {
    [key: number]: Record<'Name' | 'Name_de' | 'Name_fr' | 'Name_ja', string>;
  };
};

const fetchXivapi = async () => {
  const url = 'https://xivapi.com/Pet?columns=' + localeKeys.join(',');
  const response = await fetch(url);
  const json = (await response.json()) as XivapiResult;
  return json.Results;
};

const normalize = (
  content: XivapiResult['Results'],
): Record<'en' | 'de' | 'fr' | 'ja', string[]> => {
  const result: Record<'en' | 'de' | 'fr' | 'ja', string[]> = {
    // cactbot-ignore-missing-translations
    en: [],
    de: [],
    fr: [],
    ja: [],
  };
  for (const [, data] of Object.entries(content)) {
    // skip empty names and duplicates
    if (data.Name === '' || result.en.includes(data.Name))
      continue;

    result.en.push(data.Name);
    result.de.push(data.Name_de);
    result.fr.push(data.Name_fr);
    result.ja.push(data.Name_ja);
  }
  return result;
};

export default async (): Promise<void> => {
  const tables = {
    ...normalize(await fetchXivapi()),
    cn: Object.keys(await getCnTable('Pet', keys)).filter((k) => k !== ''),
    ko: Object.keys(await getKoTable('Pet', keys)).filter((k) => k !== ''),
  };

  const writer = new CoinachWriter(null, true);
  const header = `import { Lang } from './languages';

type PetData = {
  [name in Lang]: readonly string[];
};
`;
  await writer.writeTypeScript(
    path.join('resources', _OUTPUT_FILE),
    'gen_pet_names.ts',
    header,
    'PetData',
    false,
    tables,
  );
};
